pkg/cmd/openshift-tests/disruption/watch-endpointslice/endpointslice_controller.go

```diff
@@ -0,0 +1,272 @@
+
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	discoveryinformers "k8s.io/client-go/informers/discovery/v1"
+	discoverylisters "k8s.io/client-go/listers/discovery/v1"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/controller"
+)
+
+type EndpointSliceController struct {
+	backendPrefix string
+	serviceName   string
+	recorder      monitorapi.RecorderWriter
+	outFile       io.Writer
+
+	endpointSliceLister  discoverylisters.EndpointSliceLister
+	endpointSlicesSynced cache.InformerSynced
+
+	watcherLock sync.Mutex
+	watchers    map[string]*watcher
+
+	syncHandler func(ctx context.Context, key string) error
+	queue       workqueue.RateLimitingInterface
+}
+
+type watcher struct {
+	address                 string
+	port                    string
+	nodeName                string
+	newConnectionSampler    *backenddisruption.BackendSampler
+	reusedConnectionSampler *backenddisruption.BackendSampler
+}
+
+func NewEndpointWatcher(
+	backendPrefix string,
+	serviceName string,
+	recorder monitorapi.RecorderWriter,
+	outFile io.Writer,
+
+	endpointSliceInformer discoveryinformers.EndpointSliceInformer,
+) *EndpointSliceController {
+	c := &EndpointSliceController{
+		backendPrefix: backendPrefix,
+		serviceName:   serviceName,
+		recorder:      recorder,
+		outFile:       outFile,
+
+		endpointSliceLister:  endpointSliceInformer.Lister(),
+		endpointSlicesSynced: endpointSliceInformer.Informer().HasSynced,
+
+		watchers: map[string]*watcher{},
+
+		queue: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "EndpointWatcher"),
+	}
+	c.syncHandler = c.syncEndpointSlice
+
+	endpointSliceInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addEndpointSlice,
+		UpdateFunc: c.updateEndpointSlice,
+		DeleteFunc: c.deleteEndpointSlice,
+	})
+	return c
+}
+
+func (c *EndpointSliceController) syncEndpointSlice(ctx context.Context, key string) error {
+	namespace, name, err := cache.SplitMetaNamespaceKey(key)
+	if err != nil {
+		return err
+	}
+	endpointSlice, err := c.endpointSliceLister.EndpointSlices(namespace).Get(name)
+	if errors.IsNotFound(err) {
+		c.removeAllWatchers()
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	watchersForCurrEndpoints := map[string]*watcher{}
+	port := ""
+	for _, currPort := range endpointSlice.Ports {
+		if currPort.Port == nil {
+			continue
+		}
+		port = fmt.Sprintf("%d", *currPort.Port)
+	}
+	for _, endpoint := range endpointSlice.Endpoints {
+		for _, address := range endpoint.Addresses {
+			watchersForCurrEndpoints[address] = &watcher{
+				address: address,
+				port:    port,
+			}
+			if endpoint.NodeName != nil {
+				watchersForCurrEndpoints[address].nodeName = *endpoint.NodeName
+			}
+		}
+	}
+
+	c.watcherLock.Lock()
+	defer c.watcherLock.Unlock()
+
+	watchersToDelete := map[string]*watcher{}
```

Comment:
Lets add a comment here that this is to clean out any watchers we were holding onto that have disappeared in the current state of the EndpointSlice